java-checks/src/main/java/org/sonar/java/checks/PrintfFailCheck.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.Sets;
+import org.sonar.check.Rule;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewArrayTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+
+@Rule(key = "S2275")
+public class PrintfFailCheck extends AbstractPrintfChecker {
+
+  private static final Set<String> TIME_CONVERSIONS = Sets.newHashSet(
+    "H", "I", "k", "l", "M", "S", "L", "N", "p", "z", "Z", "s", "Q",
+    "B", "b", "h", "A", "a", "C", "Y", "y", "j", "m", "d", "e",
+    "R", "T", "r", "D", "F", "c"
+  );
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    ExpressionTree formatStringTree;
+    List<ExpressionTree> args;
+    boolean isMessageFormat = MESSAGE_FORMAT.matches(mit);
+    if (isMessageFormat && !mit.symbol().isStatic()) {
+      // only consider the static method
+      return;
+    }
+    // Check type of first argument:
+    if (mit.arguments().get(0).symbolType().is("java.lang.String")) {
```

Comment:
This tends to be overcomplicated for what we want to achieve. And it will introduce too much coupling between the checks for now. I'll stick with the duplication.